### PR TITLE
Cleanup and document fundep plugin

### PR DESCRIPTION
**Pull request by isovector**
I made the code style agree with the rest of the codebase, and I wrote down everything I know about how this works.

### CODE REVIEW

File: polysemy-plugin/src/Polysemy/Plugin/Fundep/Stuff.hs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+instance CanLookup Class where
+  lookupStrategy = tcLookupClass
+
+instance CanLookup TyCon where
+  lookupStrategy = tcLookupTyCon
+
+
+doLookup :: CanLookup a => ThingOf 'Locations a -> TcPluginM (ThingOf 'Things a)
+doLookup (mdname, name) = do
+  md  <- lookupModule (mkModuleName mdname) $ fsLit "polysemy"
```
Comment (TheMatten) [inline polysemy-plugin/src/Polysemy/Plugin/Fundep/Stuff.hs]: Some space sneaked in here :stuck_out_tongue:

File: polysemy-plugin/src/Polysemy/Plugin/Fundep.hs [hunk 2]
```diff
@@ -33,211 +33,99 @@
...
-lookupEverything :: PolysemyStuff 'Locations -> TcPluginM (PolysemyStuff 'Things)
-lookupEverything (PolysemyStuff a b c d) =
-  PolysemyStuff <$> doLookup a
-                <*> doLookup b
-                <*> doLookup c
-                <*> doLookup d
 
+------------------------------------------------------------------------------
+-- | Given a list of constraints, filter out the 'FindConstraint's.
+getFindConstraints :: PolysemyStuff 'Things -> [Ct] -> [FindConstraint]
```
Comment (isovector) [inline polysemy-plugin/src/Polysemy/Plugin/Fundep.hs]: Previously named "allMonadEffectConstraints"

File: polysemy-plugin/src/Polysemy/Plugin/Fundep.hs [hunk 3]
```diff
@@ -246,68 +134,91 @@ mustUnify (InterpreterUse b) = b
...
+  guard $ stuck == ifStuckTyCon stuff
+  (idx, [_, r, _]) <- splitTyConApp_list expr
+  guard $ idx == indexOfTyCon stuff
+  guard $ elem @[] (OrdType r) $ coerce bogus
+  pure (error "bogus proof for stuck type family", ct)
+
+
+------------------------------------------------------------------------------
+-- | Determine if there is exactly one wanted find for the @r@ in question.
+mustItUnify
```
Comment (isovector) [inline polysemy-plugin/src/Polysemy/Plugin/Fundep.hs]: Bad name!

**Comment by isovector** [comment]
@TheMatten please go through the new documentation carefully (especially for `solveFundep` and `canUnifyRecursive`) --- let me know if they make sense.

**Comment by TheMatten** [comment]
Ok, will do.

**Comment by TheMatten** [comment]
https://github.com/polysemy-research/polysemy/blob/2858c7bd84edcb23ecac530d30b6a1b92a7fbf3b/polysemy-plugin/src/Polysemy/Plugin/Fundep.hs#L29-L30
I wonder if it's still heavily based :smile:

**Pull request merged**